contrib/fetch-offline-data/instana/fetch_instana_infrastructure_metrics.py

```diff
@@ -0,0 +1,302 @@
+        params = {
+            "start": s_time,  # Convert to milliseconds
+            "end": e_time,  # Convert to milliseconds
+        }
+
+        headers = {
+            "Authorization": f"apiToken {token}",
+            "Content-Type": "application/json"
+        }
+        body = {
+            "query": "",
+            "metrics": metric_ids,
+            "snapshotIds": snapshots,
+            "plugin": plugin,
+            'rollup': granularity,
+            "timeFrame": {"to": params["end"],
+                          "windowSize": params["end"] - params["start"]}
+        }
+        parameters = {'offline': True}
+        response = requests.post(instana_api_url, params=parameters, headers=headers, json=body, verify=False)
+        if response.status_code == 200:
+            result.append(response.json()["items"])
+        else:
+            logging.error("Error fetching metrics:", response.text)
+            continue
+        s_time = e_time
+
+    return result   
+
+
+def fetch_instana_metrics(url, token, start_time, end_time, plugin, granularity=1, filter_query={}, folder = "data"):
+    time_series_metrics = []
+    metrics_count = 0
+    plugins_catalog = fetch_instana_plugins_catalog(url, token[0])
+    if plugins_catalog is None:
+        return None
+
+    plugins = list(map(lambda item: item['plugin'], plugins_catalog))
+    if plugin != None and plugin not in plugins:
+        logging.error(f"Plugin with name {plugin} not available.")
+        return None
+
+    if plugin != None:
+        plugins = [plugin]
+
+    # for each plugin, fetch snapshots and metrics
+    for p in range(0, len(plugins)):
+        plugin = plugins[p]
+        logging.info(f"fetching metrics for plugin: {plugin} [{p}/{len(plugins)}]")
+
+        # fetch the snapshots
+        snapshots = fetch_instana_snapshots(url, token[0], plugin, start_time, end_time)
+        if not snapshots:
+            continue
+
+        # fetch the metric ids
+        metrics_catalog = fetch_instana_metrics_catalog(url, token[0], plugin)
+        if metrics_catalog is None:
+            continue
+
+        # print(json.dumps(metrics_catalog))
+        # return
+
+        metric_ids = list(map(lambda item: item['metricId'], metrics_catalog))
+
+        logging.info(f"Applying filter : {filter_query}")
+
+        if filter_query != {}:
+            filtered_metric_ids = filter_metric_ids(metric_ids, filter_query)
+            metric_ids = filtered_metric_ids
+
+        # fetch the time_series metrics
+        metrics_api_limit = 5  # <<<--- limitation of the API to max 5 metrics per call
+        snapshots_api_limit = 30  # <<<--- limitation of the API to max 30 snapshots per call
+        # tokens = ["tMd8pw4WS6SdLkx1QI2KeQ", "tmdOVgSMS7ucB2rCOXkvLg", "y7l5tD1wR9G581LYOfYxKQ"]
+        counter = 0
+        for i in range(0, len(snapshots), snapshots_api_limit):
+            snapshots_subset = snapshots[i:i + snapshots_api_limit]
```

Comment:
If `len(snapshots)` is not divisible by `snapshots_api_limit`, do we have a problem of going out of bounds here?